src/main/scala/firrtl/passes/wiring/WiringUtils.scala

```diff
@@ -0,0 +1,164 @@
+        s
+      case s => s map getChildren(mname)
+    }
+    c.modules.foreach{ m =>
+      childrenMap(m.name) = Nil
+      m map getChildren(m.name)
+    }
+    childrenMap
+  }
+
+  /** Counts the number of instances of a module declared under a top module
+    */
+  def countInstances(childrenMap: ChildrenMap, top: String, module: String): Int = {
+    if(top == module) 1
+    else childrenMap(top).foldLeft(0){case (count, (i, child)) =>
+      count + countInstances(childrenMap, child, module)
+    }
+  }
+
+  /** Returns a module's lineage, containing all children lineages as well
+    */
+  def getLineage(childrenMap: ChildrenMap, module: String): Lineage =
+    Lineage(module, childrenMap(module) map (c => (c._1, getLineage(childrenMap, c._2))))
+
+  /** Sets the sink, sinkParent, source, and sourceParent fields of every
+    *  Lineage in tree
+    */
+  def setFields(sinks: Set[String], source: String)(lin: Lineage): Lineage = lin map setFields(sinks, source) match {
+    case l if sinks.contains(l.name) => l.copy(sink = true)
+    case l => 
+      val src = l.name == source
+      val sinkParent = l.children.foldLeft(false){(b, c) => b || c._2.sink || c._2.sinkParent}
+      val sourceParent = if(src) true else l.children.foldLeft(false){(b, c) => b || c._2.source || c._2.sourceParent}
+      l.copy(sinkParent=sinkParent, sourceParent=sourceParent, source=src)
+  }
+
+  /** Sets the sharedParent of lineage top
+    */
+  def setSharedParent(top: String)(lin: Lineage): Lineage = lin map setSharedParent(top) match {
+    case l if l.name == top => l.copy(sharedParent = true)
+    case l => l
+  }
+
+  /** Sets the addPort and cons fields of the lineage tree
+    */
+  def setThings(portNames:Map[String, String], compName: String)(lin: Lineage): Lineage = {
+    val funs = Seq(
+      ((l: Lineage) => l map setThings(portNames, compName)),
+      ((l: Lineage) => l match {
+        case Lineage(name, _, _, _, _, _, true, _, _) => //SharedParent
+          l.copy(addPort=Some((portNames(name), DecWire)))
+        case Lineage(name, _, _, _, true, _, _, _, _) => //SourceParent
+          l.copy(addPort=Some((portNames(name), DecOutput)))
+        case Lineage(name, _, _, _, _, true, _, _, _) => //SinkParent
+          l.copy(addPort=Some((portNames(name), DecInput)))
+        case Lineage(name, _, _, true, _, _, _, _, _) => //Sink
+          l.copy(addPort=Some((portNames(name), DecInput)))
+        case l => l
+      }),
+      ((l: Lineage) => l match {
+        case Lineage(name, _, true, _, _, _, _, _, _) => //Source
+          val tos = Seq(s"${portNames(name)}")
+          val from = compName
+          l.copy(cons=l.cons ++ tos.map(t => (t, from)))
+        case Lineage(name, _, _, _, true, _, _, _, _) => //SourceParent
+          val tos = Seq(s"${portNames(name)}")
+          val from = l.children.filter{case (i, c) => c.sourceParent}.map{case (i, c) => s"$i.${portNames(c.name)}"}.head
```

Comment:
Now that I look at this again, it can be:

```scala
val from = (l.children collect { case (i, c) if c.sourceParent => s"$i.${portNames(c.name)}" }).head
```

Not much shorter but it does only have 1 anonymous function in it instead of 2